File: app/meals.py
"""食事の一覧と投票窓。現在時刻は引数で受け取り、データベースには触れない。"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, timedelta

from . import config


@dataclass(frozen=True, order=True)
class Meal:
    day: date
    kind: str


def meal_in_period(kind: str, day: date) -> bool:
    first, last = config.PERIOD[("meal", kind)]
    return first <= day <= last


def all_meals() -> list[Meal]:
    """期間内の全食事を時系列で返す。締切の順と一致する。"""
    first = min(config.PERIOD[("meal", k)][0] for k in config.MEALS)
    last = max(config.PERIOD[("meal", k)][1] for k in config.MEALS)
    out: list[Meal] = []
    day = first
    while day <= last:
        for kind in config.MEALS:
            if meal_in_period(kind, day):
                out.append(Meal(day, kind))
        day += timedelta(days=1)
    return out


def vote_deadline(meal: Meal) -> datetime:
    shift, close_hour = config.VOTE_DEADLINE[meal.kind]
    base = meal.day + timedelta(days=shift)
    return datetime(base.year, base.month, base.day, close_hour, tzinfo=config.TZ)


def vote_state(meal: Meal, now: datetime) -> str:
    """受付開始は設けない。締切まではいつでも挙手できる。"""
    return "open" if now < vote_deadline(meal) else "closed"


def visible_meals(now: datetime) -> list[Meal]:
    """締切が未来の先頭 3 件と、直近で締め切った 1 件。"""
    meals = all_meals()
    closed = [m for m in meals if vote_state(m, now) == "closed"]
    upcoming = [m for m in meals if vote_state(m, now) != "closed"]
    return closed[-1:] + upcoming[:3]


def format_meal(meal: Meal) -> str:
    weekday = config.WEEKDAYS[meal.day.weekday()]
    return f"{meal.day.month}/{meal.day.day}({weekday}) {config.MEAL_LABELS[meal.kind]}"

File: src/meals.js
/** 食事の一覧と投票窓。現在時刻は引数で受け取り、データベースには触れない。 */

import * as config from "./config.js";
import * as jst from "./jst.js";

export function mealInPeriod(kind, day) {
  const [first, last] = config.PERIOD.meal[kind];
  return first <= day && day <= last;
}

/** 期間内の全食事を時系列で返す。締切の順と一致する。 */
export function allMeals() {
  const bounds = config.MEALS.map((kind) => config.PERIOD.meal[kind]);
  const first = bounds.map(([from]) => from).reduce((a, b) => (a < b ? a : b));
  const last = bounds.map(([, to]) => to).reduce((a, b) => (a > b ? a : b));

  const out = [];
  for (let day = first; day <= last; day = jst.addDays(day, 1)) {
    for (const kind of config.MEALS) {
      if (mealInPeriod(kind, day)) out.push({ day, kind });
    }
  }
  return out;
}

export function voteDeadline(meal) {
  const [shift, closeHour] = config.VOTE_DEADLINE[meal.kind];
  return jst.instant(jst.addDays(meal.day, shift), closeHour);
}

/** 受付開始は設けない。締切まではいつでも挙手できる。 */
export function voteState(meal, now) {
  return now < voteDeadline(meal) ? "open" : "closed";
}

/** 締切が未来の先頭 3 件と、直近で締め切った 1 件。 */
export function visibleMeals(now) {
  const meals = allMeals();
  const closed = meals.filter((meal) => voteState(meal, now) === "closed");
  const upcoming = meals.filter((meal) => voteState(meal, now) !== "closed");
  return [...closed.slice(-1), ...upcoming.slice(0, 3)];
}

export function formatMeal(meal) {
  const { month, day } = jst.parts(meal.day);
  const weekday = config.WEEKDAYS[jst.weekday(meal.day)];
  return `${month}/${day}(${weekday}) ${config.MEAL_LABELS[meal.kind]}`;
}
